#### 1. Pick one of the datasets from the ChatBot session(s) of the TUT demo (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values

In [10]:
import pandas as pd
url = "https://raw.githubusercontent.com/pointOfive/STA130_F23/main/Data/amazonbooks.csv"
df = pd.read_csv(url, encoding="ISO-8859-1")
df

Title               Author  \
0    1,001 Facts that Will Scare the S#*t Out of Yo...          Cary McNeal   
1    21: Bringing Down the House - Movie Tie-In: Th...          Ben Mezrich   
2         100 Best-Loved Poems (Dover Thrift Editions)                Smith   
3              1421: The Year China Discovered America        Gavin Menzies   
4      1493: Uncovering the New World Columbus Created      Charles C. Mann   
..                                                 ...                  ...   
320                            Where the Sidewalk Ends     Shel Silverstein   
321                                    White Privilege  Paula S. Rothenberg   
322                                Why I wore lipstick        Geralyn Lucas   
323  Worlds Together, Worlds Apart: A History of th...        Robert Tignor   
324                                  Wuthering Heights         Emily Bronte   

     List Price  Amazon Price Hard_or_Paper  NumPages               Publisher  \
0         12.95          5.18             P     304.0             Adams Media   
1         15.00         10.20             P     273.0              Free Press   
2          1.50          1.50             P      96.0      Dover Publications   
3         15.99         10.87             P     672.0        Harper Perennial   
4         30.50         16.77             P     720.0                   Knopf   
..          ...           ...           ...       ...                     ...   
320       18.99         12.24             H     192.0           HarperCollins   
321       27.55         27.55             P     160.0        Worth Publishers   
322       12.95          5.18             P     224.0     St Martin's Griffin   
323       97.50         97.50             P     480.0  W. W. Norton & Company   
324       16.99          4.95             P     344.0              CreatSpace   

     Pub year     ISBN-10  Height  Width  Thick  Weight_oz  
0      2010.0  1605506249     7.8    5.5    0.8       11.2  
1      2008.0  1416564195     8.4    5.5    0.7        7.2  
2      1995.0   486285537     8.3    5.2    0.3        4.0  
3      2008.0    61564893     8.8    6.0    1.6       28.8  
4      2011.0   307265722     8.0    5.2    1.4       22.4  
..        ...         ...     ...    ...    ...        ...  
320    2004.0    60572345     9.3    6.6    1.1       24.0  
321    2011.0  1429233443     9.1    6.1    0.7        8.0  
322    2005.0  031233446X     8.0    5.4    0.7        6.4  
323    2010.0   393934942    10.7    8.9    0.9       14.4  
324    2011.0  1463533411     7.8    5.1    1.0       16.0  

[325 rows x 13 columns]

In [12]:
missing_values = df.isnull().sum()
print(missing_values)

Title            0
Author           1
List Price       1
Amazon Price     0
Hard_or_Paper    0
NumPages         2
Publisher        1
Pub year         1
ISBN-10          0
Height           4
Width            5
Thick            1
Weight_oz        9
dtype: int64


#### 2. Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a `pandas` DataFrame has, and then

1. use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,  
2. write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot to explain these terms in the context of your dataset<br>

In [14]:
df.shape

(325, 13)

Firstly, observations represent rows in the data, while variables represent columns in the data. I think observation is a specific number, a concrete representation of data in a dataset, while variables classify this data into different categories. In this data, observations is there are 325 books in amazonbooks, the variable is "Title, Author, List Price, Amazon Price, Hard_or_Paper, NumPages, Publisher, Pub year, ISBN-10, Height, Width, Thick and Weight_oz".        

#### 3. Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset<br>

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 325 entries, 0 to 324
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Title          325 non-null    object 
 1   Author         324 non-null    object 
 2   List Price     324 non-null    float64
 3   Amazon Price   325 non-null    float64
 4   Hard_or_Paper  325 non-null    object 
 5   NumPages       323 non-null    float64
 6   Publisher      324 non-null    object 
 7   Pub year       324 non-null    float64
 8   ISBN-10        325 non-null    object 
 9   Height         321 non-null    float64
 10  Width          320 non-null    float64
 11  Thick          324 non-null    float64
 12  Weight_oz      316 non-null    float64
dtypes: float64(8), object(5)
memory usage: 33.1+ KB


In [16]:
df.describe()

List Price  Amazon Price    NumPages     Pub year      Height  \
count  324.000000    325.000000  323.000000   324.000000  321.000000   
mean    18.579753     13.333846  335.857585  2002.206790    8.163240   
std     14.252829     13.727679  161.984389    10.629002    0.918739   
min      1.500000      0.770000   24.000000  1936.000000    5.100000   
25%     13.950000      8.600000  208.000000  1998.000000    7.900000   
50%     15.000000     10.200000  320.000000  2005.000000    8.100000   
75%     19.950000     13.130000  416.000000  2010.000000    8.500000   
max    139.950000    139.950000  896.000000  2011.000000   12.100000   

            Width       Thick   Weight_oz  
count  320.000000  324.000000  316.000000  
mean     5.585000    0.907716   12.487975  
std      0.874057    0.368625    6.644648  
min      4.100000    0.100000    1.200000  
25%      5.200000    0.600000    7.800000  
50%      5.400000    0.900000   11.200000  
75%      5.900000    1.100000   16.000000  
max      9.500000    2.100000   35.200000

In [17]:
df.describe(include='all')

Title        Author  List Price  Amazon Price  \
count                325           324  324.000000    325.000000   
unique               315           255         NaN           NaN   
top     The Great Gatsby  Jodi Picoult         NaN           NaN   
freq                   3             7         NaN           NaN   
mean                 NaN           NaN   18.579753     13.333846   
std                  NaN           NaN   14.252829     13.727679   
min                  NaN           NaN    1.500000      0.770000   
25%                  NaN           NaN   13.950000      8.600000   
50%                  NaN           NaN   15.000000     10.200000   
75%                  NaN           NaN   19.950000     13.130000   
max                  NaN           NaN  139.950000    139.950000   

       Hard_or_Paper    NumPages Publisher     Pub year    ISBN-10  \
count            325  323.000000       324   324.000000        325   
unique             2         NaN       161          NaN        322   
top                P         NaN   Vintage          NaN  743273567   
freq             236         NaN        37          NaN          2   
mean             NaN  335.857585       NaN  2002.206790        NaN   
std              NaN  161.984389       NaN    10.629002        NaN   
min              NaN   24.000000       NaN  1936.000000        NaN   
25%              NaN  208.000000       NaN  1998.000000        NaN   
50%              NaN  320.000000       NaN  2005.000000        NaN   
75%              NaN  416.000000       NaN  2010.000000        NaN   
max              NaN  896.000000       NaN  2011.000000        NaN   

            Height       Width       Thick   Weight_oz  
count   321.000000  320.000000  324.000000  316.000000  
unique         NaN         NaN         NaN         NaN  
top            NaN         NaN         NaN         NaN  
freq           NaN         NaN         NaN         NaN  
mean      8.163240    5.585000    0.907716   12.487975  
std       0.918739    0.874057    0.368625    6.644648  
min       5.100000    4.100000    0.100000    1.200000  
25%       7.900000    5.200000    0.600000    7.800000  
50%       8.100000    5.400000    0.900000   11.200000  
75%       8.500000    5.900000    1.100000   16.000000  
max      12.100000    9.500000    2.100000   35.200000

In [19]:
df['List Price'].value_counts()

List Price
15.00    36
16.00    21
14.95    20
14.00    13
15.95    12
         ..
25.95     1
25.05     1
13.29     1
13.83     1
97.50     1
Name: count, Length: 99, dtype: int64

In [26]:
df["List Price"].mean()

18.579753086419753

#### 4. If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by `df.shape` and what is reported by `df.describe()` with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column<br>

I can draw the following conclusion from my presentation on the third question. "df.shape" only displays the numbers of all rows and columns of the data, and includes missing values. "df.describe()" only analyzes numeric variables in the data and ignores missing values when calculating descriptive statistics. For example, if a certain sequence has missing values, "df.describe()" will display fewer columns than "df.shape", and the difference between them is the missing value.

#### 5. Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference

- an "attribute", such as `df.shape` which does For example, if a certain sequence has missing values, df.describe will display fewer columns than df.scape, and the difference between them is the missing value. not end with `()`
- and a "method", such as `df.describe()` which does end with `()` 

"df.shape" is "attribute",don not need the"()"."df.describe()" is method, need the "()" end with.

#### 6. The `df.describe()` method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics<br>


Count: The number of non-missing (non-NaN) values in the column.<br>
Mean: The average value of the data in the column.<br>
Std: A measure of the spread or dispersion of the data from the mean. A higher standard deviation indicates that the values are more spread out from the mean.<br>
Min: The minimum or smallest value in the column.<br>
25%: The value below which 25% of the data falls. This is also called the first quartile (Q1).<br>
50%: The value below which 50% of the data falls. This is also called the second quartile (Q2).<br>
75%: The value below which 75% of the data falls. This is also called the third quartile (Q3).<br>
Max: The maximum or largest value in the column.<br>

#### 7. Missing data can be considered "across rows" or "down columns".  Consider how `df.dropna()` or `del df['col']` should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words

1. Provide an example of a "use case" in which using `df.dropna()` might be peferred over using `del df['col']`<br><br>

In [2]:
import pandas as pd

data = {'Name': ['Alice', 'Bob', 'Charlie', 'David'],
        'Age': [25, None, 30, 40],
        'Score': [85, 90, None, 75]}

abc = pd.DataFrame(data)
print(abc)


      Name   Age  Score
0    Alice  25.0   85.0
1      Bob   NaN   90.0
2  Charlie  30.0    NaN
3    David  40.0   75.0


In [4]:
abc.dropna()

Name   Age  Score
0  Alice  25.0   85.0
3  David  40.0   75.0

In this data, if we use "df.dropna()", we will only delete missing values, but if we use "del df['col']", we will delete an entire column of data. Therefore, it can be seen that using d in this situation can reduce data loss.

2. Provide an example of "the opposite use case" in which using `del df['col']` might be preferred over using `df.dropna()` <br><br>

In [11]:
import pandas as pd


data = {'Name': ['Alice', 'Bob', 'Charlie', 'David'],
        'Age': [25, 30, 35, 40],
        'Gender': ['F', 'M', 'M', 'M'],
        'Address': [None, None, None, None]} 
bbb = pd.DataFrame(data)
print(bbb)

      Name  Age Gender Address
0    Alice   25      F    None
1      Bob   30      M    None
2  Charlie   35      M    None
3    David   40      M    None


In [13]:
del bbb["Address"]
print(bbb)

      Name  Age Gender
0    Alice   25      F
1      Bob   30      M
2  Charlie   35      M
3    David   40      M


In this data, we do not have missing values, only useless values named None. If we use "bbb.dropna()", this does not remove any useless values, but if we use "del bbb['col']", we can precisely remove the useless values in the "Address" column.

3. Discuss why applying `del df['col']` before `df.dropna()` when both are used together could be important<br><br>

If a set of data contains a large number of missing values, using d directly may result in the loss of a significant amount of valid data. From this, we can conclude that we should first use "del bbb["Address"](del df['col'])" to precisely delete irrelevant columns, and then use "bbb.dropna()(df.dropna())" to delete the remaining missing values.

4. Remove all missing data from one of the datasets you're considering using some combination of `del df['col']` and/or `df.dropna()` and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.<br><br>

In [15]:
import pandas as pd

data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Age': [25, None, 35, 40],
    'Gender': ['F', 'M', None, 'M'],
    'Address': [None, None, None, None] 
}

jjj = pd.DataFrame(data)
print(jjj)


      Name   Age Gender Address
0    Alice  25.0      F    None
1      Bob   NaN      M    None
2  Charlie  35.0   None    None
3    David  40.0      M    None


In [16]:
del jjj['Address']
print(jjj)

      Name   Age Gender
0    Alice  25.0      F
1      Bob   NaN      M
2  Charlie  35.0   None
3    David  40.0      M


In [18]:
jjj.dropna()

Name   Age Gender
0  Alice  25.0      F
3  David  40.0      M

We first delete the address column through del "jjj['col']" because it is invalid data, and then use "jjj.dropna()" to delete missing values, which can avoid accidentally deleting data and make every row of our data meaningful. We didn't store any useful information under the "Address" column. This is because the "Address" column is invalid, we don't need it, deleting it allows us to keep more meaningful rows. 2. Because other columns are important, we need complete rows without missing values, so we used "jjj.dropna()" to delete the remaining rows with missing values.

#### 8. Give brief explanations in your own words for any requested answers to the questions below

In [21]:
import pandas as pd
url1 = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
ab = pd.read_csv(url1, encoding="ISO-8859-1")
ab

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

1. Use your ChatBot session to understand what `df.groupby("col1")["col2"].describe()` does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you

In [22]:
ab.groupby("survived")["age"].describe()

count       mean        std   min   25%   50%   75%   max
survived                                                           
0         424.0  30.626179  14.172110  1.00  21.0  28.0  39.0  74.0
1         290.0  28.343690  14.950952  0.42  19.0  28.0  36.0  80.0

This is to group "survived" and "age" together and combine them for analysis such as count, mean, std, min, etc.

2. Assuming you've not yet removed missing values in the manner of question "7" above, `df.describe()` would have different values in the `count` value for different data columns depending on the missingness present in the original data.  Why do these capture something fundamentally different from the values in the `count` that result from doing something like `df.groupby("col1")["col2"].describe()`?

"df.describe()" calculation is for ungrouped data, which is the total count of complete data. "df.groupby("col1")["col2"].describe()" calculates the total count of each group after grouping, and the total count of each group in "df.groupby("col1")["col2"].describe()"  is equal to the count of ungrouped data in "df.describe()".

3. Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT<br><br>
    
    1. Forget to include `import pandas as pd` in your code 
       <br> 

In [2]:
url1 = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
ab = pd.read_csv(url1, encoding="ISO-8859-1")
ab


NameError: name 'pd' is not defined

The error NameError: name 'pd' is not defined means that the pandas library (which is typically shortened to pd when imported) has not been imported in your code. You can resolve this by importing pandas at the start of your script with the following line:

2. Mistype "titanic.csv" as "titanics.csv"
       <br> 

In [2]:
import pandas as pd
url1 = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
ab = pd.read_csv(url1, encoding="ISO-8859-1")
ab

HTTPError: HTTP Error 404: Not Found

The error HTTPError: HTTP Error 404: Not Found means that the URL you're trying to access doesn't exist on the server. This could be caused by several issues:

Incorrect URL: There might be a typo or an error in the URL you're trying to access.<br>
Resource Moved: The page or resource you're requesting might have been moved or deleted.<br>
Server Issues: The server might be down or experiencing other problems that prevent access to the resource.<br>

In [1]:
import pandas as pd
url1 = "htps://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
ab = pd.read_csv(url1, encoding="ISO-8859-1")
ab

ValueError: Protocol not known: htps

In [1]:
import pandas as pd
url1 = "htps://raw.githubusercontent.com/mwasko/seaborn-data/master/titanics.csv"
ab = pd.read_csv(url1, encoding="ISO-8859-1")
ab

ValueError: Protocol not known: htps

3. Try to use a dataframe before it's been assigned into the variable
       <br> 

In [2]:
import pandas as pd
url1 = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
ab = pd.read_csv(url1, encoding="ISO-8859-1")
ab
AB.groupby("survived")["age"].describe()

NameError: name 'AB' is not defined

4. Forget one of the parentheses somewhere the code
       <br>

In [3]:
import pandas as pd
url1 = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
ab = pd.read_csv(url1, encoding="ISO-8859-1"
ab

SyntaxError: '(' was never closed (4110393104.py, line 3)

 5. Mistype one of the names of the chained functions with the code 
       <br>

In [5]:
import pandas as pd
url1 = "htps://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
ab = pd.read_csv(url1, encoding="ISO-8859-1")
ab
ab.group_by("survived")["age"].describe()

ValueError: Protocol not known: htps

6. Use a column name that's not in your data for the `groupby` and column selection 
       <br>

In [6]:
import pandas as pd
url1 = "htps://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
ab = pd.read_csv(url1, encoding="ISO-8859-1")
ab
ab.groupby("Survived")["age"].describe()

ValueError: Protocol not known: htps

  7. Forget to put the column name as a string in quotes for the `groupby` and column selection, and see if the ChatBot and google are still as helpful as they were for the previous question<br>

In [10]:
import pandas as pd
url1 = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
ab = pd.read_csv(url1, encoding="ISO-8859-1")
ab

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [11]:
 ab.groupby(sex)["age"].describe()

NameError: name 'sex' is not defined

In [13]:
ab.groupby("sex")[age].describe()

NameError: name 'age' is not defined

I prefer ChatGPT because ChatGPT can interact with me and provide precise answers to my questions, while Google requires me to filter out information that is relevant to me, which I believe wastes a lot of my valuable time. In my opinion, ChatGPT is more efficient and enables us to better self-study knowledge.

#### 9. Have you reviewed the course [wiki-textbook](https://github.com/pointOfive/stat130chat130/wiki) and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?<br>

YES<br>
Chat Gpt link:https://chatgpt.com/share/34e0e8a6-8f7f-4ca2-b678-0cb5c385e015<br>
https://chatgpt.com/share/a65dfd0e-eaff-45f1-b10f-c7b72bb69374<br>